server.c

```diff
@@ -0,0 +1,300 @@
+
+    struct config_t *config;
+};
+
+struct params_t
+{
+    struct srv_context_t *context;
+    int socket_fd;
+};
+
+static int
+send_task(const int client_sfd, struct task_t *task, bool *result)
+{
+    int status;
+
+    status = sendall(client_sfd, task, sizeof(struct task_t), 0);
+    if (status == -1) return -1;
+
+    int size;
+    status = recvall(client_sfd, &size, sizeof(size), 0);
+    if (status == -1) return -1;
+
+    if (size != 0)
+    {
+        status = recv(client_sfd, task->password, size, 0);
+        if (status == -1) return -1;
+    }
+
+    *result = (size != 0);
+    return 0;
+}
+
+static void *
+serve_client(void *arg)
+{
+    struct params_t *params = (struct params_t *) arg;
+    struct srv_context_t *context = (struct srv_context_t *) params->context;
+    int client_sfd = params->socket_fd;
+    free(params);
+
+    while (true)
+    {
+        struct task_t task;
+        queue_pop(&context->queue, &task);
+
+        task.to = task.from;
+        task.from = 0;
+        bool found = false;
+        int status = send_task(client_sfd, &task, &found);
+        if (status == -1)
+        {
+            queue_push(&context->queue, &task);
+            break;
+        }
+        if (found)
+        {
+            memcpy(context->password, task.password, sizeof(task.password));
+            context->found = true;
+            context->done = true;
+        }
+
+        pthread_mutex_lock(&context->tasks_mutex);
+        --context->tasks_running;
+        pthread_mutex_unlock(&context->tasks_mutex);
+
+        if (context->tasks_running == 0 || context->found)
+            pthread_cond_signal(&context->tasks_cond);
+    }
+
+    pthread_mutex_lock(&context->set_mutex);
+    set_remove_sock(&context->set, client_sfd);
+    close(client_sfd);
+    pthread_mutex_unlock(&context->set_mutex);
+
+    return NULL;
+}
+
+static bool
+srv_password_handler(void *context, struct task_t *task)
+{
+    struct srv_context_t *ctx = (struct srv_context_t *) context;
+
+    pthread_mutex_lock(&ctx->tasks_mutex);
+    ++ctx->tasks_running;
+    pthread_mutex_unlock(&ctx->tasks_mutex);
+
+    queue_push(&ctx->queue, task);
+    return (ctx->password[0] != 0);
+}
+
+static void *
+srv_server(void *arg)
+{
+    struct params_t *params = (struct params_t *) arg;
+    struct srv_context_t *context = (struct srv_context_t *) params->context;
+    int server_sfd = params->socket_fd;
+
+    while (true)
+    {
+        // Printing to stderr to be able to check output in tests
+        fprintf(stderr, "Waiting for new connections...\n");
+        int client_socket = accept(server_sfd, NULL, NULL);
+        if (client_socket == -1)
+            handle_error("accept");
+        fprintf(stderr, "Got new connection...\n");
+
+        pthread_mutex_lock(&context->set_mutex);
+        struct node_t *node = set_take_last(&context->set);
+        node->socket_fd = client_socket;
+
+        struct params_t *params = malloc(sizeof(struct params_t));
```

Comment:
Emmm... too early for dynamic alloc.
All you need is to ensure that thread started synchronously - just one more synchronisation on thread start. You also need to handle failure of pthread_create